dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/task/TaskSetState.java

```diff
@@ -16,60 +16,134 @@
  */
 package com.google.edwmigration.dumper.application.dumper.task;
 
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.edwmigration.dumper.application.dumper.task.TaskCategory.REQUIRED;
+import static com.google.edwmigration.dumper.application.dumper.task.TaskState.FAILED;
+import static java.util.stream.Collectors.counting;
+import static java.util.stream.Collectors.groupingBy;
+
+import com.google.common.collect.ImmutableList;
 import java.util.HashMap;
 import java.util.Map;
 import javax.annotation.CheckForNull;
 import javax.annotation.Nonnull;
+import javax.annotation.ParametersAreNonnullByDefault;
 import javax.annotation.concurrent.GuardedBy;
 import net.jcip.annotations.ThreadSafe;
 
 /** @author shevek */
 public interface TaskSetState {
 
+  @ParametersAreNonnullByDefault
+  static class TaskResultSummary {
+    private final String exception;
+    private final Task<?> task;
+    private final TaskState state;
+
+    private TaskResultSummary(String exception, Task<?> task, TaskState state) {
+      this.exception = exception;
+      this.task = task;
+      this.state = state;
+    }
+
+    @Nonnull
+    static TaskResultSummary create(Task<?> task, TaskResult<?> result) {
+      String optionalException = String.format(": %s", result.getException());
+      String exceptionDescription = result.getException() == null ? "" : optionalException;
+      return new TaskResultSummary(exceptionDescription, task, result.getState());
+    }
+
+    @Nonnull
+    @Override
+    public String toString() {
+      return String.format("Task %s (%s) %s%s", state, task.getCategory(), task, exception);
+    }
+  }
+
+  @ParametersAreNonnullByDefault
+  public static class TasksReport {
+
+    private final long count;
+    private final TaskState state;
+
+    @Nonnull
+    @Override
+    public String toString() {
+      return String.format("%d TASKS %s", count, state);
+    }
+
+    TasksReport(long count, TaskState state) {
+      this.count = count;
+      this.state = state;
+    }
+  }
+
   @ThreadSafe
   public static class Impl implements TaskSetState {
 
-    @GuardedBy("lock")
+    @GuardedBy("this")
```

Comment:
One should never expose one's mutexes, they should always be private objects. Exposing a mutex makes it non-private, which means someone else can lock on it, and cause deadlock.

Reply:
The hypothetical is that two threads access some `TaskSetState` instance `x` simultaneously. One enters a block, which is `synchronized (x) {...`, while the other calls `x`'s member method, whose implementation requires a mutex.

The suggestion is that concurrent access should be allowed for these threads. The PR opts to enforce the mutex.

It's hard to be sure about the intention of someone writing `synchronized (x) {...` in this context, but it's unlikely that they would expect a different thread operating on critical sections of `x` at the same time through a different mutex.

In the deadlock example, replacing the call to a `synchronized` method on `x` with a `synchronized (x) {...` block leads to the same behavior. The risk is inherent to `synchronized` blocks which depend on some shared object, adding the method doesn't increase it.